bootstrap/kubeadm/api/v1alpha3/kubeadmbootstrapconfig_types.go

```diff
@@ -172,7 +172,31 @@ type File struct {
 	Encoding Encoding `json:"encoding,omitempty"`
 
 	// Content is the actual content of the file.
-	Content string `json:"content"`
+	Content string `json:"content,omitempty"`
+
+	// ContentFrom is a referenced source of content to populate the file.
+	ContentFrom *FileSource `json:"contentFrom,omitempty"`
+}
+
+// FileSource is a union of all possible external source types for file data.
+// Only one field may be populated in any given instance. Developers adding new
+// sources of data for target systems should add them here.
+type FileSource struct {
+	// Secret represents a secret that should populate this file.
+	// +optional
+	Secret *SecretFileSource `json:"secret,omitempty"`
```

Comment:
This shouldn't be optional I think? If one specifies a `FileSource` in `file.contentFrom: {}` it should fail validation saying that `secret` is required

Reply:
Since we only have one I was thinking if in the future we change this from required to optional it'd be breaking? But I think you mentioned today it's fine, and I concur that makes sense. Will make required.

The webhook does validate the stricter scenario already, I think you're correct regardless.